# Notebook 3 : Metadata-Enhanced Retrieval

In the previous notebook, we saw how we can evaluate our RAG application. In this notebook, we'll see how we can use metadata filtering to improve the quality of our retrieval system.

We'll do so in 3 parts

1. First we'll show how we can use a VLM like GPT-4o to generate metadata for our items
2. Then we'll see how we can use this metadata to filter the items we retrieve
3. Finally, we'll use this new metadata filtering and quantify the improvement in our retrieval system with regards to recall and MRR.

Let's get started!



## Using VLMs to generate Metadata




## Using Metadata for Filtering


What we want is a way for us to map a user query to a set of metadata fields that we'd like to use to filter out items. We've defined a `taxonomy.yml` ahead of time that defines the metdata fields which we've used to annotate our dataset that we previously ingested.

What we'll do is to read in the taxonomy and make sure that our generated metadata matches this taxonomy. 

In [5]:
from helpers.taxonomy import process_taxonomy_file

taxonomy_map = process_taxonomy_file("./taxonomy.yml")

In [6]:
taxonomy_map.keys()

dict_keys(['Tops', 'Bottoms', 'Dresses', 'Outerwear'])

In [7]:
taxonomy_map["Tops"]

{'product_type': ['T-Shirts',
  'Blouses',
  'Sweaters',
  'Cardigans',
  'Tank Tops',
  'Hoodies',
  'Sweatshirts'],
 'attributes': {'Sleeve Length': ['Sleeveless',
   'Short Sleeve',
   '3/4 Sleeve',
   'Long Sleeve'],
  'Neckline': ['Crew Neck', 'V-Neck', 'Turtleneck', 'Scoop Neck', 'Cowl Neck'],
  'Fit': ['Regular', 'Slim', 'Oversized', 'Cropped']}}

In [20]:
from typing import Optional
from pydantic import BaseModel, model_validator, ValidationInfo


class Attribute(BaseModel):
    name: str
    values: list[str]


class QueryFilters(BaseModel):
    attributes: list[Attribute]
    min_price: Optional[float] = None
    max_price: Optional[float] = None
    category: str
    product_type: list[str]

    @model_validator(mode="after")
    def validate_attributes(self, info: ValidationInfo):
        taxonomy_data = info.context["taxonomy_data"]
        # Validate category exists in taxonomy
        if self.category not in taxonomy_data:
            raise ValueError(
                f"Invalid category: {self.category}. Valid categories are {taxonomy_data.keys()}"
            )

        # Validate product types
        valid_types = taxonomy_data[self.category]["product_type"]
        for product_type in self.product_type:
            if product_type not in valid_types:
                raise ValueError(
                    f"Invalid product type: {product_type}. Valid product types are {valid_types}"
                )

        # Validate attribute exists in taxonomy
        valid_attrs = taxonomy_data[self.category]["attributes"]
        for attr in self.attributes:
            if attr.name not in valid_attrs:
                raise ValueError(f"Invalid attribute name: {attr.name}")
            for value in attr.values:
                if value not in valid_attrs[attr.name]:
                    raise ValueError(
                        f"Invalid value {value} for attribute {attr.name}. Valid values are {valid_attrs[attr.name]}"
                    )

        return self

We'll then use this to make a LLM call to generate these metadata fields for our queries

In [60]:
from openai import OpenAI
import instructor
from helpers.taxonomy import process_taxonomy_file
from rich import print

client = instructor.from_openai(OpenAI())
taxonomy_data = process_taxonomy_file("taxonomy.yml")
query = "I want a Tank-Top that's got a short sleeve or sleeveless which is under 100 bucks for an interview"


resp = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {
            "role": "system",
            "content": """
          You are a helpful assistant that extracts user requirements from a query.
                    
Use these references:
- Taxonomy: {{ taxonomy }}

Guidelines:
- If a filter isn't needed, return an empty list
- Only add attributes and filters that a user has mentioned explicitly
- Only use values from the provided taxonomy. 
- If the attribute exists on multiple types, make sure that you only look at the specific types listed under the category you have chosen
- If the user hasn't mentioned a specific product type, lets just use all of them
- if the user gives a range (Eg. around 50), just give a buffer of 20 on each side (Eg. 30-70)
- if the user gives a vague price (Eg. I have a high budget), just set max price to 1000
- only classify an item as unisex if the user has explicitly mentioned it and default to Women's categories by default.
- If you're looking at blouses, make sure to include tank tops along the way and vice versa
- if the user mentions user bottoms and doesn't specify a specific length - let's include both short and long bottoms such as jeans, shorts and pants
- make sure to look carefully at the user's query to determine if they've specified a specific fit - eg. regular, relaxed, cropped. ( Relaxed and Relaxed should always go together)


Extract the requirements and format them according to the QueryFilters model.
            """,
        },
        {"role": "user", "content": query},
    ],
    context={
        "taxonomy_data": taxonomy_data,
    },
    response_model=QueryFilters,
)

print(resp)

QueryFilters(
    attributes=[Attribute(name='Sleeve Length', values=['Sleeveless'])],
    min_price=None,
    max_price=100.0,
    category='Tops',
    product_type=['Tank Tops']
)

We can then see a few more other sentences to make sure that the filters are working as expected.

In [34]:
import instructor
from openai import AsyncOpenAI


async def extract_query_filters(
    client: instructor.AsyncInstructor, query: str
) -> QueryFilters:
    """
    Extract structured filters from a natural language query using LLM.

    Args:
        query (str): Natural language query from user

    Returns:
        QueryFilters: Structured filters extracted from the query
    """
    return await client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[
            {
                "role": "system",
                "content": "You are a helpful assistant that extracts user requirements from a query. Refer to this taxonomy for valid categories, subcategories, product types and attributes: {{ taxonomy_data }}. If a filter isn't needed, just return an empty list. If the user is looking for a specific attribute, just return the attribute name and the values that the user is looking for",
            },
            {"role": "user", "content": query},
        ],
        context={
            "taxonomy_data": taxonomy_data,
        },
        response_model=QueryFilters,
    )


client = instructor.from_openai(AsyncOpenAI())
print(await extract_query_filters(client, "Need a new skirt around 40 bucks"))

QueryFilters(attributes=[], min_price=0.0, max_price=40.0, category='Bottoms', product_type=['Skirts'])

Let's now see how we can use this metadata filtering with LanceDB.

In [120]:
import json
from lancedb import connect
import pandas as pd


def retrieve_and_filter(query: str, table, filters: QueryFilters, max_k=75):
    query_parts = []

    # We do a prefilter on category,price and material since these will always be provided
    query_parts.append(f"category='{filters.category}'")

    if filters.min_price:
        query_parts.append(f"price >= {filters.min_price}")
    if filters.max_price:
        query_parts.append(f"price <= {filters.max_price}")

    query_string = " AND ".join(query_parts)
    items = (
        table.search(query=query)
        .where(query_string, prefilter=True)
        .limit(max_k)
        .to_list()
    )

    items = [
        {
            **item,
            "attributes": json.loads(item["attributes"]),
        }
        for item in items
    ]

    if filters.product_type:
        items = [item for item in items if item["product_type"] in filters.product_type]

    if filters.attributes:
        for attr in filters.attributes:
            if not attr.values:
                continue
            curr_items = []
            for item in items:
                attr_name = attr.name
                attr_values = attr.values
                item_attr_values = item["attributes"]
                for item_attr in item_attr_values:
                    if (
                        item_attr["name"] == attr_name
                        and item_attr["value"] in attr_values
                    ):
                        curr_items.append(item)
                        break

            items = curr_items

    return items

In [ ]:
test_query = "I want a skirt that is at most 300 bucks"
client = instructor.from_openai(AsyncOpenAI())
db = connect("./lancedb")
table = db.open_table("items")
generated_filter = await extract_query_filters(client, test_query)
results = retrieve_and_filter(test_query, table, generated_filter)
pd.DataFrame(results).loc[
    :,
    [
        "title",
        "description",
        "category",
        "product_type",
        "price",
        "attributes",
        "in_stock",
    ],
]


title  ... in_stock
0  High-Waisted Pencil Skirt  ...     True
1    White Eyelet Mini Skirt  ...     True
2         Plaid Pencil Skirt  ...    False
3          Black Denim Skirt  ...     True
4     Green Plaid Mini Skirt  ...     True

[5 rows x 7 columns]

We can see that the filter is working as expected. Let's now apply this new filtering to our retrieval system and see how it performs.

In [72]:
import json

# Load in queries that we generated previously
with open("queries.json", "r") as f:
    queries = [json.loads(line) for line in f]


In [73]:
import instructor
from lancedb.table import Table
from tqdm.asyncio import tqdm_asyncio


async def retrieve_item(client: instructor.AsyncInstructor, item: dict, table: Table):
    generated_filter = await extract_query_filters(client, item["query"])
    results = retrieve_and_filter(item["query"], table, generated_filter)
    return {
        "query": item["query"],
        "retrieved_items": results,
        "expected_items": [item["id"]],
        "filters": generated_filter,
    }


client = instructor.from_openai(AsyncOpenAI())
coros = [retrieve_item(client, query, table) for query in queries]
results = await tqdm_asyncio.gather(*coros)


100%|██████████| 38/38 [01:48<00:00,  2.86s/it]


We'll now use this to evaluate the performance of our retrieval system

In [67]:
retrieved_items = [
    [item["id"] for item in result["retrieved_items"]] for result in results
]
labels = [item["expected_items"] for item in results]

In [74]:
from helpers.metrics import get_metrics_at_k

metrics = get_metrics_at_k(metrics=["mrr", "recall"], sizes=[5, 10, 25])
computed_metrics = [
    {
        metric: score_fn(retrieved_item_ids, desired_item)
        for metric, score_fn in metrics.items()
    }
    for retrieved_item_ids, desired_item in zip(retrieved_items, labels)
]
df = pd.DataFrame(computed_metrics)

In [75]:
df.mean()

mrr@5        0.500000
mrr@10       0.527778
mrr@25       0.527778
recall@5     0.500000
recall@10    0.750000
recall@25    0.750000
dtype: float64

### Braintrust

When working with these larger evaluation datasets, it's important to have a way to visualise the results easily. For this, we'll switch over to using Braintrust to visualise these results so that we can iterate on the prompts for our metadata filters better.

We'll redefine our `extract_query_filters` function below again and iteratively modify our prompt to get better results.

In [130]:
import instructor


async def extract_query_filters(
    client: instructor.AsyncInstructor, query: str
) -> QueryFilters:
    return await client.chat.completions.create(
        model="gpt-4o",
        messages=[
            {
                "role": "system",
                "content": """
You are a helpful assistant that extracts user requirements from a query.

Use this following taxonomy as a reference for what fields are available to you. Only use values from the provided taxonomy.

<taxonomy>
{{ taxonomy_data }}
</taxonomy>

Guidelines:

- If a specific filter isn't needed, just return an empty list or null value for that
- If the attribute exists on multiple types, make sure that you only look at the specific types listed under the category you have chosen
- Make sure that you've chosen from the right attribute values for each attribute type. This is very important.

Here are some general rules about how to generate these filters
1. Dresses and Skirts should always go together
2. Potential Filters should only be added if the user has explicitly mentioned it. When selecting filter values, aim to make them more flexible. For instance, if the user is asking for a well fitting top, we can consider both regular and relaxed fit. If another attribute value might be a good match for the user's query, include it too. 
3. If the user hasn't mentioned the attribute for the product type in his query, don't include that attribute in the filter. For instance if the user only mentions that they want something that's comfortable, don't include a filter for the fit of the product.
4. If the user mentions a rough range ( eg. around 50 bucks), let's just use a buffer of 30 bucks on each side ( Eg. 20-80)
5. If the user mentions a vague price (Eg. I have a high budget), just set max price to 1000
6. Make sure to look carefully at the user's query to determine if they've specified a specific fit - eg. regular, relaxed, cropped. ( Relaxed and Relaxed should always go together)

""",
            },
            {"role": "user", "content": query},
        ],
        context={
            "taxonomy_data": taxonomy_data,
        },
        response_model=QueryFilters,
    )


Now let's define our braintrust `AsyncEval` and then use it to iterate on our prompt.

In [ ]:
from braintrust import EvalAsync, Score
from helpers.metrics import get_metrics_at_k
import lancedb
import openai


def evaluate_braintrust(input, output, **kwargs):
    metrics = get_metrics_at_k(metrics=["mrr", "recall"], sizes=[1, 3, 5, 10, 15, 25])
    return [
        Score(
            name=metric,
            score=score_fn(output, kwargs["expected"]),
            metadata={"query": input, "result": output, **kwargs["metadata"]},
        )
        for metric, score_fn in metrics.items()
    ]


client = instructor.from_openai(openai.AsyncOpenAI())
taxonomy_data = process_taxonomy_file("taxonomy.yml")
with open("queries.json", "r") as f:
    queries = [json.loads(line) for line in f]
db = lancedb.connect("./lancedb")
table = db.open_table("items")


async def generate_filters_and_retrieve_items(query: dict, hooks) -> dict:
    generated_filter = await extract_query_filters(client, query["query"])
    results = retrieve_and_filter(query["query"], table, generated_filter)
    items_without_vector = [
        {k: v for k, v in item.items() if k != "vector"} for item in results
    ]

    hooks.meta(filters=generated_filter.model_dump(), items=items_without_vector)
    return [item["id"] for item in results]


await EvalAsync(
    "query-generation",
    data=lambda: [{"input": query, "expected": [query["id"]]} for query in queries],
    task=generate_filters_and_retrieve_items,
    scores=[evaluate_braintrust],
)

Skipping git metadata. This is likely because the repository has not been published to a remote yet. Remote named 'origin' didn't exist
Experiment main-1739390828 is running at https://www.braintrust.dev/app/567/p/query-generation/experiments/main-1739390828
query-generation (data): 38it [00:00, 74792.84it/s]


query-generation (tasks):   0%|          | 0/38 [00:00<?, ?it/s]

/var/folders/ws/q_m6c6qs3n553603dk_zvrgc0000gn/T/ipykernel_7946/737981449.py:32: DeprecationWarning: meta() is deprecated. Use the metadata field directly instead.
  hooks.meta(



=========================SUMMARY=========================
main-1739390828 compared to main-1739390737:
73.68% (-01.99%) 'mrr@1'     score	(1 improvements, 1 regressions)
78.07% (-02.11%) 'mrr@3'     score	(1 improvements, 1 regressions)
78.07% (-02.11%) 'mrr@5'     score	(1 improvements, 1 regressions)
79.18% (-02.14%) 'mrr@10'    score	(1 improvements, 1 regressions)
79.18% (-02.14%) 'mrr@15'    score	(1 improvements, 1 regressions)
79.18% (-02.14%) 'mrr@25'    score	(1 improvements, 1 regressions)
73.68% (-01.99%) 'recall@1'  score	(1 improvements, 1 regressions)
84.21% (-02.28%) 'recall@3'  score	(1 improvements, 1 regressions)
84.21% (-02.28%) 'recall@5'  score	(1 improvements, 1 regressions)
92.11% (-02.49%) 'recall@10' score	(1 improvements, 1 regressions)
92.11% (-02.49%) 'recall@15' score	(1 improvements, 1 regressions)
92.11% (-02.49%) 'recall@25' score	(1 improvements, 1 regressions)

1739390828.04s start
1739390845.33s end
11.27s (+287.71%) 'duration'	(11 improvements, 27 r

EvalResultWithSummary(summary="...", results=[...])

We can see that the filters are working as expected. By implementing metadata filters, we've been able to improve the recall and MRR of our retrieval system by a significant margin as seen below.

| Metric | Semantic Search (Baseline) | Semantic Search + Metadata Filters |
|---------|--------------------------|----------------------------------|
| MRR@5   | 0.47 | 0.78 (+65.43%) |
| MRR@10  | 0.49 | 0.79 (+61.67%) |
| MRR@15  | 0.49 | 0.79 (+61.06%) |
| MRR@20  | 0.49 | 0.79 (+60.05%) |
| MRR@25  | 0.49 | 0.79 (+60.05%) |
| Recall@5  | 0.63 | 0.84 (+33.33%) |
| Recall@10 | 0.76 | 0.92 (+20.70%) |
| Recall@15 | 0.79 | 0.92 (+16.67%) |
| Recall@20 | 0.84 | 0.92 (+9.38%) |
| Recall@25 | 0.84 | 0.92 (+9.38%) |

Most importantly, we've been able to quantify the improvement in our retrieval at each step. This allows us to know the impact of each step in our retrieval pipeline. 

## Better Captioning

We can also use the same concept to explore other ways to improve our retrieval system. For instance, we can improve the captioning of our items by providing more information to be embedded. 

Let's see how we can do this below.



In [132]:
from lancedb.pydantic import LanceModel, Vector
import lancedb
from lancedb.embeddings import get_registry

from datasets import load_dataset

dataset = load_dataset("ivanleomk/ai-engineer-summit-ecommerce-taxonomy")["train"]
func = get_registry().get("openai").create(name="text-embedding-3-small")


# Define a Model that will be used as the schema for our collection
class Item(LanceModel):
    id: int
    title: str
    description: str = func.SourceField()
    brand: str
    category: str
    product_type: str
    attributes: str
    material: str
    pattern: str
    price: float
    vector: Vector(func.ndims()) = func.VectorField()
    in_stock: bool


db = lancedb.connect("./lancedb")
table_name = "items_better_captioning"

if table_name not in db.table_names():
    table = db.create_table(table_name, schema=Item, mode="overwrite")
    entries = []
    for row in dataset:
        entries.append(
            {
                "id": row["id"],
                "title": row["title"],
                "description": f"""
title: {row["title"]}
description: {row["description"]}
brand: {row["brand"]}
category: {row["category"]}
product_type: {row["product_type"]}
price: {row["price"]}
attributes: {json.dumps(row["attributes"])}
""".strip(),
                "brand": row["brand"],
                "category": row["category"],
                "product_type": row["product_type"],
                "attributes": row["attributes"],
                "material": row["material"],
                "pattern": row["pattern"],
                "price": row["price"],
                "in_stock": row["in_stock"],
            }
        )

    table.add(entries)

table = db.open_table(table_name)
print(f"{table.count_rows()} rows in the table")

[2025-02-12T20:18:05Z WARN  lance::dataset::write::insert] No existing dataset at /Users/ivanleo/Documents/coding/ai-engineer-summit/lancedb/items_better_captioning.lance, it will be created


191 rows in the table

In [135]:
from braintrust import EvalAsync, Score
from helpers.metrics import get_metrics_at_k
import lancedb
import openai


def evaluate_braintrust(input, output, **kwargs):
    metrics = get_metrics_at_k(metrics=["mrr", "recall"], sizes=[1, 3, 5, 10, 15, 25])
    return [
        Score(
            name=metric,
            score=score_fn(output, kwargs["expected"]),
            metadata={"query": input, "result": output, **kwargs["metadata"]},
        )
        for metric, score_fn in metrics.items()
    ]


table = db.open_table("items_better_captioning")


def task(query: dict, hooks) -> dict:
    return [item["id"] for item in table.search(query["query"]).limit(25).to_list()]


await EvalAsync(
    "query-generation",
    data=lambda: [{"input": query, "expected": [query["id"]]} for query in queries],
    task=task,
    scores=[evaluate_braintrust],
)

Skipping git metadata. This is likely because the repository has not been published to a remote yet. Remote named 'origin' didn't exist
Experiment main-1739391749 is running at https://www.braintrust.dev/app/567/p/query-generation/experiments/main-1739391749
query-generation (data): 38it [00:00, 24200.36it/s]


query-generation (tasks):   0%|          | 0/38 [00:00<?, ?it/s]


=========================SUMMARY=========================
main-1739391749 compared to main-1739391719:
50.00% 'mrr@1'     score
53.95% 'mrr@3'     score
56.18% 'mrr@5'     score
58.00% 'mrr@10'    score
58.22% 'mrr@15'    score
58.48% 'mrr@25'    score
50.00% 'recall@1'  score
60.53% 'recall@3'  score
71.05% 'recall@5'  score
84.21% 'recall@10' score
86.84% 'recall@15' score
92.11% 'recall@25' score

1739391749.20s start
1739391749.95s end
0.60s (+26.48%) 'duration'	(7 improvements, 31 regressions)

See results for main-1739391749 at https://www.braintrust.dev/app/567/p/query-generation/experiments/main-1739391749


EvalResultWithSummary(summary="...", results=[...])

In this case, we see an average increase of ~18% for MRR and ~10% for recall across all values of `k` for this new improved description that we're embedding.

| Metric | Semantic Search (Baseline) | Semantic Search + Metadata |
|---------|--------------------------|----------------------------------|
| MRR@5   | 0.47 | 0.56 (+19.15%) |
| MRR@10  | 0.49 | 0.58 (+18.43%) |
| MRR@15  | 0.49 | 0.58 (+18.42%) |
| MRR@20  | 0.49 | 0.58 (+18.21%) |
| MRR@25  | 0.49 | 0.58 (+18.21%) |
| Recall@5  | 0.63 | 0.71 (+12.50%) |
| Recall@10 | 0.76 | 0.84 (+10.34%) |
| Recall@15 | 0.79 | 0.87 (+10.00%) |
| Recall@20 | 0.84 | 0.92 (+9.52%) |
| Recall@25 | 0.84 | 0.92 (+9.52%) |

This combined with our original metadata filtering approach would significantly improve the results that we would get. Most importantly, with these synthetic datasets that we've created, we're able to know and quantify the impact at every step of the way. This is huge in helping us to really systematically improve our RAG application 